pkg/api/registered/registered.go

```diff
@@ -25,71 +26,97 @@ import (
-					version, os.Getenv("KUBE_API_VERSIONS"), validGroupVersions)
+				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s.",
+					version, os.Getenv("KUBE_API_VERSIONS"))
 			}
+			envRequestedVersions[gv] = struct{}{}
+		}
+	}
+}
 
-			// Verify that the version is valid.
-			valid, ok := validGroupVersions[gv]
-			if !ok || !valid {
-				// Not a valid API version.
-				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s. List of valid API versions: %v",
-					version, os.Getenv("KUBE_API_VERSIONS"), validGroupVersions)
-			}
+// RegisterVersions add the versions to the registeredVersions.
+func RegisterVersions(versions ...unversioned.GroupVersion) {
+	for _, v := range versions {
+		registeredVersions[v] = struct{}{}
+	}
+}
 
-			supportedVersions = append(supportedVersions, gv)
-		}
+// IsEnvReuqestedVersion returns if the version is requested via the
+// KUBE_API_VERSIONS environment variable. If the environment variable is empty,
+// then it always returns true.
+func IsEnvReuqestedVersion(v unversioned.GroupVersion) bool {
+	if len(envRequestedVersions) == 0 {
+		return true
 	}
+	_, found := envRequestedVersions[v]
+	return found
+}
 
-	RegisteredGroupVersions = supportedVersions
+// IsRegisteredVersion returns if a version is enabled.
+func IsRegisteredVersion(v unversioned.GroupVersion) bool {
+	_, found := registeredVersions[v]
+	return found
 }
 
-// Returns true if the given api version is one of the registered api versions.
-func IsRegisteredAPIGroupVersion(gv unversioned.GroupVersion) bool {
-	for _, currGV := range RegisteredGroupVersions {
-		if currGV == gv {
-			return true
-		}
+// RegisteredVersions returns all enabled versions.
+func RegisteredVersions() (ret []unversioned.GroupVersion) {
+	for v := range registeredVersions {
+		ret = append(ret, v)
 	}
-	return false
+	return
 }
 
-// GroupVersionsForGroup returns the registered versions of a group in the form
-// of "group/version".
-func GroupVersionsForGroup(group string) []unversioned.GroupVersion {
-	ret := []unversioned.GroupVersion{}
-	for _, v := range RegisteredGroupVersions {
+// RegisteredVersionsForGroup returns all enabled versions for a group.
+func RegisteredVersionsForGroup(group string) (ret []unversioned.GroupVersion) {
+	for v := range registeredVersions {
 		if v.Group == group {
 			ret = append(ret, v)
 		}
 	}
-	return ret
+	return
+}
+
+// ValidateEnvRequestedVersions returns a list of versions that are requested in
+// the KUBE_API_VERSIONS environment variable, but not enabled.
+func ValidateEnvRequestedVersions() []unversioned.GroupVersion {
+	var missingVersions []unversioned.GroupVersion
+	for v := range envRequestedVersions {
+		if _, found := registeredVersions[v]; !found {
+			missingVersions = append(missingVersions, v)
+		}
+	}
+	return missingVersions
+}
+
+// ValidateEnvRequestedVersionsOrDie panics if KUBE_API_VERSIONS contains
+// versions that are not enabled.
+func ValidateEnvRequestedVersionsOrDie() {
```

Comment:
I'd prefer to leave the dying to calling the code.  I wouldn't expect to see the call in many spots.